charger/etrel.go

```diff
@@ -0,0 +1,252 @@
+	case 1, 2:
+		return api.StatusA, nil
+	case 3, 5, 6, 7, 9:
+		return api.StatusB, nil
+	case 4:
+		return api.StatusC, nil
+	default:
+		return api.StatusNone, fmt.Errorf("invalid status: %d", u)
+	}
+}
+
+// Enabled implements the api.Charger interface
+func (wb *Etrel) Enabled() (bool, error) {
+	b, err := wb.conn.ReadHoldingRegisters(etrelRegMaxCurrent, 2)
+	if err != nil {
+		return false, err
+	}
+
+	return encoding.Float32(b) > 0, nil
+}
+
+// Enable implements the api.Charger interface
+func (wb *Etrel) Enable(enable bool) error {
+	if enable {
+		return wb.setCurrent(wb.current)
+	}
+
+	b := make([]byte, 1)
+	binary.BigEndian.PutUint16(b, 1)
+
+	_, err := wb.conn.WriteMultipleRegisters(etrelRegStop, 1, b)
+	return err
+}
+
+func (wb *Etrel) setCurrent(current float32) error {
+	b := make([]byte, 4)
+	binary.BigEndian.PutUint32(b, math.Float32bits(wb.current))
+
+	_, err := wb.conn.WriteMultipleRegisters(etrelRegMaxCurrent, 1, b)
+	return err
+}
+
+// MaxCurrent implements the api.Charger interface
+func (wb *Etrel) MaxCurrent(current int64) error {
+	return wb.MaxCurrentMillis(float64(current))
+}
+
+var _ api.ChargerEx = (*Etrel)(nil)
+
+// MaxCurrentMilli implements the api.ChargerEx interface
+func (wb *Etrel) MaxCurrentMillis(current float64) error {
+	if current < 6 {
+		return fmt.Errorf("invalid current %.1f", current)
+	}
+
+	f := float32(current)
+
+	err := wb.setCurrent(f)
+	if err == nil {
+		wb.current = f
+	}
+
+	return err
+}
+
+var _ api.ChargeTimer = (*Etrel)(nil)
+
+// ChargingTime implements the api.ChargeTimer interface
+func (wb *Etrel) ChargingTime() (time.Duration, error) {
+	b, err := wb.conn.ReadInputRegisters(etrelRegChargeTime, 4)
+	if err != nil {
+		return 0, err
+	}
+
+	return time.Duration(binary.BigEndian.Uint64(b)) * time.Second, nil
+}
+
+var _ api.Meter = (*Etrel)(nil)
+
+// CurrentPower implements the api.Meter interface
+func (wb *Etrel) CurrentPower() (float64, error) {
+	b, err := wb.conn.ReadInputRegisters(etrelRegPower, 2)
+	if err != nil {
+		return 0, err
+	}
+
+	return float64(encoding.Float32(b) * 1e3), err
+}
+
+// var _ api.MeterEnergy = (*Etrel)(nil)
+
+// // TotalEnergy implements the api.MeterEnergy interface
+// func (wb *Etrel) TotalEnergy() (float64, error) {
+// 	b, err := wb.conn.ReadInputRegisters(etrelRegTotalEnergy, 2)
+// 	if err != nil {
+// 		return 0, err
+// 	}
+
+// 	return float64(binary.BigEndian.Uint32(b)) / 10, err
+// }
+
+var _ api.ChargeRater = (*Etrel)(nil)
+
+// ChargedEnergy implements the api.ChargeRater interface
+func (wb *Etrel) ChargedEnergy() (float64, error) {
+	b, err := wb.conn.ReadInputRegisters(etrelRegSessionEnergy, 2)
+	if err != nil {
+		return 0, err
+	}
+
+	return float64(encoding.Float32(b)), err
+}
+
+var _ api.MeterCurrent = (*Etrel)(nil)
+
+// Currents implements the api.MeterCurrent interface
+func (wb *Etrel) Currents() (float64, float64, float64, error) {
+	var currents []float64
+	for _, regCurrent := range etrelRegCurrents {
```

Comment:
Würde ich eher in 6 Register auf einen Read umstellen statt 3 einzelne Abfragen.
Ist performanter und schöner.

Reply:
Nice. Dann müssen wir das aber überall machen. Immer schön die gleichen Patterns ;)